File: server-legacy/src/main/java/de/pnp/manager/model/character/PlayerCharacter.java
package de.pnp.manager.model.character;

import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import de.pnp.manager.model.Battle;
import de.pnp.manager.model.Currency;
import de.pnp.manager.model.ICurrency;
import de.pnp.manager.model.loot.ILootTable;
import de.pnp.manager.model.loot.LootTable;
import javafx.beans.property.*;

@JsonSerialize(as = IPlayerCharacter.class)
public class PlayerCharacter extends PnPCharacter implements IPlayerCharacter {

    protected StringProperty race;
    protected StringProperty age;
    protected IntegerProperty experience;
    protected StringProperty profession;
    protected ObjectProperty<ICurrency> currency;
    protected StringProperty history;
    protected StringProperty gender;

    public PlayerCharacter(String characterID, Battle battle) {
        this(characterID, battle, new LootTable());
    }

    public PlayerCharacter(String characterID, Battle battle, ILootTable lootTable) {
        super(characterID, battle, lootTable);
        this.race = new SimpleStringProperty("");
        this.age = new SimpleStringProperty("");
        this.experience = new SimpleIntegerProperty(0);
        this.profession = new SimpleStringProperty("");
        this.currency = new SimpleObjectProperty<>(new Currency(0));
        this.history = new SimpleStringProperty("");
        this.gender = new SimpleStringProperty("");
    }

    @Override
    public ILootTable getFinishedLootTable() {
        ILootTable result = super.getFinishedLootTable();
        result.add(getCurrency());

        return result;
    }

    @Override
    public String getRace() {
        return race.get();
    }

    @Override
    public String getGender() {
        return gender.get();
    }

    public void setGender(String gender) {
        this.gender.set(gender);
    }

    public StringProperty raceProperty() {
        return race;
    }

    public void setRace(String race) {
        this.race.set(race);
    }

    @Override
    public String getAge() {
        return age.get();
    }

    public StringProperty ageProperty() {
        return age;
    }

    public void setAge(String age) {
        this.age.set(age);
    }

    @Override
    public int getExperience() {
        return experience.get();
    }

    public IntegerProperty experienceProperty() {
        return experience;
    }

    public void setExperience(int experience) {
        this.experience.set(experience);
    }

    @Override
    public String getProfession() {
        return profession.get();
    }

    public StringProperty professionProperty() {
        return profession;
    }

    public void setProfession(String profession) {
        this.profession.set(profession);
    }

    @Override
    public ICurrency getCurrency() {
        return currency.get();
    }

    public ObjectProperty<ICurrency> currencyProperty() {
        return currency;
    }

    public void setCurrency(ICurrency currency) {
        this.currency.set(currency);
    }

    @Override
    public String getHistory() {
        return history.get();
    }

    public StringProperty historyProperty() {
        return history;
    }

    public void setHistory(String history) {
        this.history.set(history);
    }
}

File: app/src/main/java/de/pnp/manager/app/model/PlayerCharacter.kt
package de.pnp.manager.app.model

import de.pnp.manager.app.state.ConfigHelper
import de.pnp.manager.model.ICurrency
import de.pnp.manager.model.character.IInventory
import de.pnp.manager.model.character.IPlayerCharacter
import de.pnp.manager.model.character.data.IArmorPosition
import de.pnp.manager.model.character.data.IPrimaryAttribute
import de.pnp.manager.model.character.data.ISecondaryAttribute
import de.pnp.manager.model.character.state.IMemberState
import de.pnp.manager.model.item.IArmor
import de.pnp.manager.model.item.IJewellery
import de.pnp.manager.model.item.IWeapon
import de.pnp.manager.model.loot.ILootTable
import de.pnp.manager.model.other.ISpell
import de.pnp.manager.model.other.ITalent

class PlayerCharacter : IPlayerCharacter, PnPCharacter() {
    private var race: String = ""
    private var gender: String = ""
    private var age: String = ""
    private var experience: Int = 0
    private var profession: String = ""
    private var currency: Currency? = null
    private var history: String = ""

    fun getExperiencePercentage(): Int {
        val modifier = ConfigHelper.getInt("character.experienceLevelModifier")
        val neededXP = this.level * modifier
        return (experience / neededXP) * 100
    }

    fun getExperiencePercentageString() : String {
        val modifier = ConfigHelper.getInt("character.experienceLevelModifier")
        val neededXP = this.level * modifier
        return "$experience/$neededXP"
    }

    override fun getRace(): String {
        return race
    }

    fun setRace(race: String) {
        this.race = race
    }

    override fun getGender(): String {
        return gender
    }

    fun setGender(gender: String) {
        this.gender = gender
    }

    override fun getAge(): String {
        return age
    }

    fun setAge(age: String) {
        this.age = age
    }

    override fun getExperience(): Int {
        return experience
    }

    fun setExperience(experience: Int) {
        this.experience = experience
    }

    override fun getProfession(): String {
        return profession
    }

    fun setProfession(profession: String) {
        this.profession = profession
    }

    override fun getCurrency(): Currency? {
        return currency
    }

    fun setCurrency(currency: Currency) {
        println("currency: " + currency.coinValue)
        this.currency=  currency
    }

    override fun getHistory(): String {
        return history
    }

    fun setHistory(history: String) {
        this.history = history
    }

}